src/rdb/rdb_raft.c

```diff
@@ -1119,54 +1119,139 @@ rdb_raft_update_node(struct rdb *db, uint64_t index, raft_entry_t *entry, rdb_vo
+	if (entry->type == RAFT_LOGTYPE_NORMAL) {
+		int rc;
+
+		rc = rdb_tx_count_vops(db, entry->data.buf, entry->data.len);
+		if (rc < 0)
+			return rc;
+		count += rc;
+	} else if (raft_entry_is_cfg_change(entry)) {
+		count += RDB_RAFT_UPDATE_NODE_NVOPS;
+	} else {
+		D_ERROR(DF_DB ": unknown entry type %d\n", DP_DB(db), entry->type);
+		return -DER_IO;
+	}
+
+	/* Count those that will be invoked when storing the entry. */
+	count += RDB_RAFT_ENTRY_NVOPS;
+
+	return count;
+}
+
 /*
  * Must invoke no more than RDB_RAFT_ENTRY_NVOPS VOS TX operations directly
  * (i.e., not including those invoked by rdb_tx_apply and
- * rdb_raft_update_node).
+ * rdb_raft_update_node) per VOS TX.
  */
 static int
-rdb_raft_log_offer_single(struct rdb *db, raft_entry_t *entry, uint64_t index, rdb_vos_tx_t vtx)
-{
-	d_iov_t			keys[2];
-	d_iov_t			values[2];
-	struct rdb_entry	header;
-	int			n = 0;
-	bool			crit;
-	int			rc;
-
-	D_ASSERTF(index == db->d_lc_record.dlr_tail, DF_U64" == "DF_U64"\n",
-		  index, db->d_lc_record.dlr_tail);
+rdb_raft_log_offer_single(struct rdb *db, raft_entry_t *entry, uint64_t index)
+{
+	rdb_vos_tx_t     vtx;
+	bool             skip_tx_apply = false;
+	d_iov_t          keys[2];
+	d_iov_t          values[2];
+	struct rdb_entry header;
+	int              n;
+	bool             crit = true;
+	bool             dirtied_tail;
+	bool             dirtied_kvss;
+	int              rc;
+
+retry:
+	/* Initialize or reset per TX variables. */
+	dirtied_tail = false;
+	dirtied_kvss = false;
+
+	/* Begin a VOS TX. */
+	if (skip_tx_apply) {
+		D_ASSERTF(entry->type == RAFT_LOGTYPE_NORMAL, "%d == %d\n", entry->type,
+			  RAFT_LOGTYPE_NORMAL);
+		rc = RDB_RAFT_ENTRY_NVOPS;
+	} else {
+		rc = rdb_raft_entry_count_vops(db, entry);
+		if (rc < 0) {
+			DL_ERROR(rc, DF_DB ": failed to count VOS operations for entry %ld",
+				 DP_DB(db), index);
+			return rc;
+		}
+	}
+	rc = rdb_vos_tx_begin(db, rc /* nvops */, &vtx);
+	if (rc != 0) {
+		DL_ERROR(rc, DF_DB ": failed to begin VOS TX for entry %ld", DP_DB(db), index);
+		return rc;
+	}
 
 	/*
-	 * If this is an rdb_tx entry, apply it. Note that the updates involved
-	 * won't become visible to queries until entry index is committed.
-	 * (Implicit queries resulted from rdb_kvs cache lookups won't happen
-	 * until the TX releases the locks for the updates after the
-	 * rdb_tx_commit() call returns.)
+	 * Update the log tail. To get the same minor epoch for all MC updates
+	 * across different VOS TXs, we update the MC first.
 	 */
+	D_ASSERTF(index == db->d_lc_record.dlr_tail, DF_U64 " == " DF_U64 "\n", index,
+		  db->d_lc_record.dlr_tail);
+	db->d_lc_record.dlr_tail = index + 1;
+	dirtied_tail = true;
+	d_iov_set(&values[0], &db->d_lc_record, sizeof(db->d_lc_record));
+	rc = rdb_mc_update(db->d_mc, RDB_MC_ATTRS, 1 /* n */, &rdb_mc_lc, &values[0], vtx);
+	if (rc != 0) {
+		DL_ERROR(rc, DF_DB ": failed to update log tail " DF_U64, DP_DB(db),
+			 db->d_lc_record.dlr_tail);
+		goto out_vtx;
+	}
+
```

Comment:
[Note] This corresponds to "reorder the LC and MC updates" paragraph in the commit message; the other changes in this PR are almost all for "fix the handling of deterministic errors".